src/subject_name/dns_name.rs

```diff
@@ -130,24 +122,145 @@ impl<'a> DnsNameRef<'a> {
     }
 }
 
-/// Requires the `alloc` feature.
-#[cfg(feature = "alloc")]
 impl core::fmt::Debug for DnsNameRef<'_> {
     fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
-        let lowercase = self.clone().to_owned();
-        f.debug_tuple("DnsNameRef").field(&lowercase.0).finish()
+        f.write_str("DnsNameRef(\"")?;
+
+        // Convert each byte of the underlying ASCII string to a `char` and
+        // downcase it prior to formatting it. We avoid self.clone().to_owned()
+        // since it requires allocation.
+        for &ch in self.0 {
+            core::fmt::Display::fmt(&char::from(ch).to_ascii_lowercase(), f)?;
```

Comment:
I think we can write this as `f.write_char(..)` instead, which is a bit cleaner?

Reply:
oh, yeah, i forgot that `Formatter` had a `write_char` method, that's definitely nicer.

Reply:
Force pushed this fix. Much nicer! Ty